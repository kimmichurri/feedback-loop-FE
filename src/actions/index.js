export const hasError = (message) => ({
  type: 'HAS_ERROR',
  message
})

export const isLoading = (boolean) => ({
  type: 'IS_LOADING',
  isLoading: boolean
})

export const setSurvey = (survey) => ({
  type: 'SET_SURVEY',
  survey
})

export const setCohorts = (cohorts) => ({
  type: 'SET_COHORTS',
  cohorts
})

export const setUser = (user) => ({
  type: 'SET_USER',
  user
})

export const setCurrentCohort = (cohort) => ({
  type: 'SET_CURRENT_COHORT',
  cohort
})