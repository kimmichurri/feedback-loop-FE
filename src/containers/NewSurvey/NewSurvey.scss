.new-survey-container {
  background: $white;
  height: 500px;
  width: 80%;
  margin: auto;
  padding: 25px;
  display: grid;
  grid-template-rows: 100px 1fr 50px;
}

.new-survey-landing-form {
  display: flex;
  flex-direction: column;
  padding: 5px;
}

.survey-name {
  font-size: 34px;
  height: 51px;
  width: 100%;
  border: none;
  padding: 5px;
  color: $main-grey;
  line-height: 135%;
    &:focus {
      outline: none;
    }
}

.begin-create-survey-label {
  font-weight: bold;
  font-family: $main-font;
  margin-right: 5px;
  color: $main-grey;
  text-align: left;
  padding: 5px;
  display: flex;
}

.controls {
  display: flex;
  justify-content: flex-end;
}

.begin-new-survey-button {
  background-color: $main-blue;
  border-radius: 10px;
  border: none;
  color: $accent-teal;
  font-family: $main-font;
  font-weight: bold;
  font-size: 16px;
  height: 50px;
  width: auto;
  padding: 0px 20px;
  margin-left: 20px;
  &:hover {
    cursor: pointer;
    background-color: $main-blue-hover;
  }
}


.exp {
  margin: 0px;
  padding-right: 5px;
}

.popper {
  width: 500px;
}