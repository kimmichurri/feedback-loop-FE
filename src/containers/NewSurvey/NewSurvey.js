import React, { Component } from 'react';
import { connect } from 'react-redux'
import DatePicker from 'react-datepicker';
import "react-datepicker/dist/react-datepicker.css"
import Question from '../../components/Question/Question'
import shortid from 'shortid'
import { setSurvey } from '../../actions'
import { withRouter } from 'react-router-dom'
import cogoToast from 'cogo-toast';

export class NewSurvey extends Component {
  constructor(props) {
    super(props);
    this.state = {
      surveyName: '',
      surveyExpiration: new Date(),
      questions: [ { id: shortid()} ],
    }
  }

  handleSubmit = (e) => {
    e.preventDefault()
    if(!this.state.surveyName || !this.state.questions[0].questionTitle || !this.state.questions[0].options[0].option_1.description) {
      this.warnToast('Your survey must have a name and at least one question')
    } else {
      this.props.setSurvey(this.state)
      this.props.history.push('/recipients')
    }
  }

  warnToast = (message) => {
    cogoToast.warn(message, {position: 'bottom-left'})
  }

  handleChange = (e) => {
    const { name, value} = e.target
    this.setState({
      [name]: value
    })
  }

  handleDate = (date) => {
    this.setState({
      surveyExpiration: date
    })
  }

  updateQuestions = (newQuestion) => {
    if(this.state.questions.length) {
      const updatedQuestions = this.state.questions.map(question => {
        if (question.id === newQuestion.id) {
          question = newQuestion
        }
        return question
      })
      this.setState({
        questions: updatedQuestions
      })
    }
  }

  addQuestion = () => {
    this.setState({
      questions: [...this.state.questions, { id: shortid() }]
    })
  }

  displayQuestion = (question) => (
    <Question
      key={question.id}
      {...question}
      updateQuestions={this.updateQuestions} />
  )

  render() {

    let questionCards
    if (this.state.questions.length) {
      questionCards = this.state.questions.map(question => this.displayQuestion(question))
    }

    return(
      <div className='new-survey-container'>
        <form onSubmit={this.handleSubmit} className='new-survey-landing-form'>
          <input
            type="text"
            name="surveyName"
            className="survey-name"
            placeholder="Add Survey Name"
            value={this.state.surveyName}
            onChange={this.handleChange}
          />
          <label className="begin-create-survey-label">
            <p className="exp">Expiration Date:</p>
            <DatePicker
              selected={this.state.surveyExpiration}
              onChange={this.handleDate}
              showTimeSelect
              timeFormat="HH:mm"
              timeIntervals={30}
              dateFormat="MMMM d, yyyy h:mm aa"
              timeCaption="Time"
              popperClassName="popper"
              />
          </label>
        </form>
        <div className='question-cards'>
          {questionCards}
        </div>
        <div className='controls'>
          <button className="begin-new-survey-button" onClick={this.addQuestion}>Add a Question</button>
          <button className="begin-new-survey-button" onClick={this.handleSubmit}>Choose Recipients</button>
        </div>
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  setSurvey: (survey) => dispatch(setSurvey(survey))
})

export default withRouter(connect(null, mapDispatchToProps)(NewSurvey))
