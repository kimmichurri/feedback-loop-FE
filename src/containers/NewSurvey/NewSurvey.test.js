import React from 'react'
import { NewSurvey, mapDispatchToProps } from './NewSurvey'
import { shallow } from 'enzyme'
import * as actions from '../../actions'

describe('NewSurvey', () => {
  let wrapper
  let mockFn = jest.fn()
  let mockHistory = { push: jest.fn() }

  beforeEach(() => {
    wrapper = shallow(<NewSurvey 
      setSurvey={mockFn}
      history={mockHistory}
    />)
  })

  it.skip('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have a default state', () => {
    expect(wrapper.state().surveyName).toEqual('')
    expect(wrapper.state().questions).toEqual([])
  })

  it('should call setSurvey when handleSubmit is invoked', () => {
    const preventDefault = { preventDefault: jest.fn() }
    wrapper.instance().handleSubmit(preventDefault)
    expect(mockFn).toHaveBeenCalled()
  })

  it('should update state on change', () => {
    const mockEvent = { target: { name: 'surveyName', value: 'My Survey'}}
    wrapper.find('.survey-name').simulate('change', mockEvent)
    expect(wrapper.state('surveyName')).toBe('My Survey')
  })

  it('should update state with a new question when updateQuestions is invoked', () => {
    const mockState = {
      questions: [{ id: 1}]
    }
    const updatedQuestion = { 
      id: 1,
      questionTitle: 'My Question',
      option_1: {pointValue: 1, questionText: 'How are you feeling today?'}
    }
    const expected = [{ id: 1, questionTitle: 'My Question', option_1: {pointValue: 1, questionText: 'How are you feeling today?'}}]
    wrapper.setState(mockState)
    wrapper.instance().updateQuestions(updatedQuestion)
    expect(wrapper.state('questions')).toEqual(expected)
  })

  it('should add a question to the questions array in state when addQuestion is invoked', () => {
    const mockState = {
      questions: [{ id: 1, questionTitle: 'My Question', option_1: {pointValue: 1, questionText: 'How are you feeling today?'}}]
    }
    wrapper.setState(mockState)
    wrapper.instance().addQuestion()
    expect(wrapper.state('questions')).toHaveLength(2)
  })

  it('should call displayQuestion cards once for every question in state', () => {
    const mockState = {
      questions: [{ id: 1, questionTitle: 'My Question', option_1: {pointValue: 1, questionText: 'How are you feeling today?'}}]
    }
    const displayQuestionSpy = jest.spyOn(wrapper.instance(), 'displayQuestion')
    wrapper.setState(mockState)
    expect(displayQuestionSpy).toHaveBeenCalledTimes(1)
  })

  describe('mapDispatchToProps', () => {
    it('should set a survey in the store', () => {
      let survey = {
        surveyName: 'My Survey',
        questions: [{ id: 1, questionTitle: 'My Question', option_1: {pointValue: 1, questionText: 'How are you feeling today?'}}]
      }
      let dispatch = jest.fn()
      let actionToDispatch = actions.setSurvey(survey)
      const mappedProps = mapDispatchToProps(dispatch)
      mappedProps.setSurvey(survey)
      expect(dispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})