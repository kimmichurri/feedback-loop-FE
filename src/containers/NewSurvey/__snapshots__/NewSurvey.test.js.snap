// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewSurvey should match the snapshot 1`] = `
<div
  className="new-survey-container"
>
  <form
    className="new-survey-landing-form"
    onSubmit={[Function]}
  >
    <input
      className="survey-name"
      name="surveyName"
      onChange={[Function]}
      placeholder="Untitled Survey"
      type="text"
      value=""
    />
    <label
      className="begin-create-survey-label"
    >
      <p
        className="exp"
      >
        Expiration Date:
      </p>
      <DatePicker
        allowSameDay={false}
        dateFormat="MMMM d, yyyy h:mm aa"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        monthsShown={1}
        nextMonthButtonLabel="Next month"
        onBlur={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        popperClassName="popper"
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2019-05-22T20:05:13.533Z}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showTimeInput={false}
        showTimeSelect={true}
        strictParsing={false}
        timeCaption="Time"
        timeFormat="HH:mm"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      />
    </label>
  </form>
  <div
    className="question-cards"
  />
  <div
    className="controls"
  >
    <button
      className="begin-new-survey-button"
      onClick={[Function]}
    >
      Add a Question
    </button>
    <button
      className="begin-new-survey-button"
      onClick={[Function]}
    >
      Choose Recipients
    </button>
  </div>
</div>
`;
